proof.go

```diff
@@ -465,6 +465,201 @@ func (proof Proof) VerifyInclusion(h hash.Hash, nid namespace.ID, leavesWithoutN
+	if err := nth.ValidateNodeFormat(root); err != nil {
+		return false, fmt.Errorf("root does not match the NMT hasher's hash format: %w", err)
+	}
+	// check that all the proof.Notes() are valid w.r.t the NMT hasher
+	for _, node := range proof.Nodes() {
+		if err := nth.ValidateNodeFormat(node); err != nil {
+			return false, fmt.Errorf("proof nodes do not match the NMT hasher's hash format: %w", err)
+		}
+	}
+	// check that all the subtree roots are valid w.r.t the NMT hasher
+	for _, subtreeRoot := range subtreeRoots {
+		if err := nth.ValidateNodeFormat(subtreeRoot); err != nil {
+			return false, fmt.Errorf("inner nodes does not match the NMT hasher's hash format: %w", err)
+		}
+	}
+
+	// get the subtree roots leaf ranges
+	ranges, err := ToLeafRanges(proof.Start(), proof.End(), subtreeWidth)
+	if err != nil {
+		return false, err
+	}
+
+	// check whether the number of ranges matches the number of subtree roots.
+	// if not, make an early return.
+	if len(subtreeRoots) != len(ranges) {
+		return false, fmt.Errorf("number of subtree roots %d is different than the number of leaf ranges %d", len(subtreeRoots), len(ranges))
+	}
+
+	var computeRoot func(start, end int) ([]byte, error)
+	// computeRoot can return error iff the HashNode function fails while calculating the root
+	computeRoot = func(start, end int) ([]byte, error) {
+		// reached a leaf
+		if end-start == 1 {
+			// if the leaf index falls within the proof range, pop and return a
+			// leaf
+			if proof.Start() <= start && start < proof.End() {
+				// this check should always be the case.
+				// however, it is added to avoid nil pointer exceptions
+				if len(ranges) != 0 {
+					// advance the list of ranges
+					ranges = ranges[1:]
+				}
+				// advance leafHashes
+				return popIfNonEmpty(&subtreeRoots), nil
+			}
+
+			// if the leaf index is outside the proof range, pop and return a
+			// proof node (which in this case is a leaf) if present, else return
+			// nil because leaf doesn't exist
+			return popIfNonEmpty(&proof.nodes), nil
+		}
+
+		// if the current range does not overlap with the proof range, pop and
+		// return a proof node if present, else return nil because subtree
+		// doesn't exist
+		if end <= proof.Start() || start >= proof.End() {
+			return popIfNonEmpty(&proof.nodes), nil
+		}
+
+		if len(ranges) != 0 && ranges[0].Start == start && ranges[0].End == end {
+			ranges = ranges[1:]
+			return popIfNonEmpty(&subtreeRoots), nil
+		}
+
+		// Recursively get left and right subtree
+		k := getSplitPoint(end - start)
+		left, err := computeRoot(start, start+k)
+		if err != nil {
+			return nil, fmt.Errorf("failed to compute subtree root [%d, %d): %w", start, start+k, err)
+		}
+		right, err := computeRoot(start+k, end)
+		if err != nil {
+			return nil, fmt.Errorf("failed to compute subtree root [%d, %d): %w", start+k, end, err)
+		}
+
+		// only right leaf/subtree can be non-existent
+		if right == nil {
```

Comment:
check this: https://github.com/celestiaorg/nmt/pull/260#discussion_r1655417247

That test case goes over trees that are not a power of 2 and tests all combinations and everything passes. However, I am not sure all edge cases are covered/the test really is testing all combinations. 

If we are sure that we can handle all type of trees, then we can remove the comment. Otherwise, we can look into removing this part of code

Reply:
As I eluded previously,
>This if statement makes me think that you are considering a non-full tree, right?
If this is the case, then we need to also clarify how a proof looks like for a non-full tree. Ideally, the proof.nodes should contain a node for such an empty right subtree, where the value of the node would be a nested hash of an empty leaf (depending on the level of the node), right?

If the aim is to cover cases that the rightmost side of the tree is empty, then we first need to decide on how a proof for such a tree looks like.  There are two cases here:
- We want to allow proofs for non-full trees:
To me, even when the tree is not full, still the hash of nil values should be returned hence the can never be a nil node (i.e., nil left or right subtree), there should always be a hash value (hence we should never expect `right==nil`). 

- But if we do not want to cover non-full trees, then, again there would not be any empty right or left subtree root.


The last point: Nevertheless,  the `right`  in this `if` statement is relative, and does not necessarily mean the right most part of the tree, rather it is a right subtree of any node in the tree.

Reply:
@staheri14
Concerning this:
> Ideally, the proof.nodes should contain a node for such an empty right subtree, where the value of the node would be a nested hash of an empty leaf (depending on the level of the node), right?

Why is that?

> we need to also clarify how a proof looks like for a non-full tree

the same way it is for an RFC 6962, no? the inner nodes required to compute the root. Are there any special cases that you think would need to be specified?

> To me, even when the tree is not full, still the hash of nil values should be returned hence the can never be a nil node (i.e., nil left or right subtree), there should always be a hash value (hence we should never expect right==nil).

That's not how it works currently. In non-full trees, we return a nil when we encounter a non-existing right node. Do you mean we should change this behaviour also for the current implementation?

To sum-up, I removed the:
```
		if end-start == 1 {
			// if the leaf index falls within the proof range, pop and return a
			// leaf
			if proof.Start() <= start && start < proof.End() {
				// this check should always be the case.
				// however, it is added to avoid nil pointer exceptions
				if len(ranges) != 0 {
					// advance the list of ranges
					ranges = ranges[1:]
				}
				// advance leafHashes
				return popIfNonEmpty(&subtreeRoots), nil
			}
			// if the leaf index is outside the proof range, pop and return a
			// proof node (which in this case is a leaf) if present, else return
			// nil because leaf doesn't exist
			return popIfNonEmpty(&proof.nodes), nil
		}
```
block entirely and everything seems to work.

What else is blocking to the merging of this PR? so that we get ahead and audit the change as it's needed in Celestia-node.

Reply:
Some of the changes discussed in this thread are here: 70bc2b0 , let's merge it now and follow up if needed

Reply:
If there is something left to discuss, let's do it in separate issues 👍